source/common/upstream/upstream_connection_pool_impl.cc

```diff
@@ -0,0 +1,173 @@
+#include "common/upstream/upstream_connection_pool_impl.h"
+
+#include "common/common/lock_guard.h"
+
+namespace Envoy {
+namespace Upstream {
+
+UpstreamConnectionPoolImpl::UpstreamConnectionPoolImpl(Event::Dispatcher& dispatcher)
+    : dispatcher_(dispatcher) {}
+
+UpstreamConnectionPoolImpl::~UpstreamConnectionPoolImpl() {
+  for (auto& pool : connection_pools_) {
+    auto& clients = pool.second.clients;
+    while (!clients.empty()) {
+      clients.front()->client_connection_->close(Network::ConnectionCloseType::NoFlush);
+    }
+  }
+
+  // Make sure all clients are destroyed before we are destroyed.
+  dispatcher_.clearDeferredDeleteList();
+}
+
+void UpstreamConnectionPoolImpl::OfferConnection(HostConstSharedPtr host,
+                                                 UpstreamConnectionEssence connection_essence) {
+  auto shared_connection_essence =
+      std::make_shared<UpstreamConnectionEssence>(std::move(connection_essence));
+  dispatcher_.post([this, host, shared_connection_essence]() {
+    AcceptConnectionInDispatcherThread(host, std::move(*shared_connection_essence));
+  });
+  {
+    Thread::LockGuard lock(mu_);
+    ConnectionPool& pool = connection_pools_[host];
+    ++pool.available_connections;
```

Comment:
It's not available until the post completes, right? Shouldn't this be done in the context of the post? Or you want to account for it ahead of time?

Reply:
available_connections = clients.size() - pending posted callbacks to RetrieveConnectionInDispatcherThread

This could be more clear if I track pending client additions, pending client removals and clients.size() separately.  I would allow retrieves from the pool if:
  clients.size() + pending_adds - pending_removes > 0